// @vector-im/compound-web/dist/components/Menu/CheckboxMenuItem.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("react/jsx-runtime");
const React = require("react");
const MenuItem = require("./MenuItem.cjs");
const Checkbox = require("../Form/Controls/Checkbox/Checkbox.cjs");
const CheckboxMenuItem = React.forwardRef(
  function CheckboxMenuItem2({ className, label, onSelect, checked, disabled }, ref) {
    const toggleId = React.useId();
    const onChange = React.useCallback(() => {
    }, []);
    return /* @__PURE__ */ jsxRuntime.jsx(
      MenuItem.MenuItem,
      {
        as: "div",
        role: "menuitemcheckbox",
        "aria-checked": checked,
        className,
        label,
        onSelect,
        disabled,
        Icon: /* @__PURE__ */ jsxRuntime.jsx(
          Checkbox.CheckboxInput,
          {
            id: toggleId,
            ref,
            "aria-hidden": true,
            checked,
            disabled,
            onChange
          }
        )
      }
    );
  }
);
exports.CheckboxMenuItem = CheckboxMenuItem;
