// @vector-im/compound-web/dist/components/Menu/ToggleMenuItem.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("react/jsx-runtime");
const React = require("react");
const MenuItem = require("./MenuItem.cjs");
const Toggle = require("../Form/Controls/Toggle/Toggle.cjs");
const ToggleMenuItem = React.forwardRef(
  function ToggleMenuItem2({ className, Icon, label, onSelect, checked, disabled, onClick }, ref) {
    const toggleId = React.useId();
    const onChange = React.useCallback(() => {
    }, []);
    return /* @__PURE__ */ jsxRuntime.jsx(
      MenuItem.MenuItem,
      {
        as: "div",
        role: "menuitemcheckbox",
        "aria-checked": checked,
        className,
        Icon,
        label,
        onSelect,
        disabled,
        onClick,
        children: /* @__PURE__ */ jsxRuntime.jsx(
          Toggle.ToggleInput,
          {
            id: toggleId,
            ref,
            "aria-hidden": true,
            checked,
            disabled,
            onChange
          }
        )
      }
    );
  }
);
exports.ToggleMenuItem = ToggleMenuItem;
